File: src/teams-config.ts
import { readFileSync } from "node:fs";
import { parse } from "yaml";

export interface TeamEntry {
  readonly org: string;
  readonly slug: string;
}

export class InvalidTeamsConfig extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidTeamsConfig";
  }
}

export class TeamsConfig {
  readonly entries: readonly TeamEntry[];

  constructor(raw: unknown) {
    this.entries = Object.freeze(parseEntries(raw));
  }

  static load(path: string): TeamsConfig {
    const raw = (parse(readFileSync(path, "utf8")) as unknown) ?? [];
    return new TeamsConfig(raw);
  }
}

function parseEntries(raw: unknown): TeamEntry[] {
  if (!Array.isArray(raw)) {
    throw new InvalidTeamsConfig(`expected a list of teams, got ${describe(raw)}`);
  }

  const seen = new Set<string>();
  return raw.map((row, i) => {
    if (!isPlainObject(row)) {
      throw new InvalidTeamsConfig(`entry ${i}: expected a mapping`);
    }
    const org = requireString(row, "org", i);
    const slug = requireString(row, "team_slug", i);
    const key = `${org}/${slug}`;
    if (seen.has(key)) {
      throw new InvalidTeamsConfig(`duplicate entry: ${key}`);
    }
    seen.add(key);
    return Object.freeze({ org, slug });
  });
}

function requireString(row: Record<string, unknown>, key: string, index: number): string {
  const raw = row[key];
  const value = typeof raw === "string" ? raw.trim() : "";
  if (value.length === 0) {
    throw new InvalidTeamsConfig(`entry ${index}: missing ${key}`);
  }
  return value;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function describe(value: unknown): string {
  if (value === null) return "null";
  if (Array.isArray(value)) return "array";
  return typeof value;
}

File: lib/octladder/teams_config.rb
require "yaml"

class TeamsConfig
  Entry = Data.define(:org, :slug)

  class InvalidConfig < ArgumentError; end

  DEFAULT_PATH = File.expand_path("../../config/teams.yml", __dir__)

  def self.load(path = DEFAULT_PATH)
    new(YAML.safe_load_file(path) || [])
  end

  attr_reader :entries

  def initialize(raw)
    @entries = parse(raw).freeze
  end

  private

  def parse(raw)
    raise InvalidConfig, "expected a list of teams, got #{raw.class}" unless raw.is_a?(Array)

    seen = {}
    raw.map.with_index do |row, i|
      raise InvalidConfig, "entry #{i}: expected a mapping" unless row.is_a?(Hash)

      entry = Entry.new(
        org:  require_string(row, "org", i),
        slug: require_string(row, "team_slug", i)
      )

      key = [ entry.org, entry.slug ]
      raise InvalidConfig, "duplicate entry: #{entry.org}/#{entry.slug}" if seen[key]
      seen[key] = true

      entry
    end
  end

  def require_string(row, key, index)
    value = row[key].to_s.strip
    raise InvalidConfig, "entry #{index}: missing #{key}" if value.empty?
    value
  end
end
